common/src/lib.rs

```diff
@@ -286,21 +311,16 @@ pub fn caller_nonce_fetch_inc(caller_id: H256) -> u64 {
     original_nonce
 }
 
-pub fn insert_waiting_message(prog_id: H256, msg_id: H256, message: Message) {
-    sp_io::storage::set(&wait_key(prog_id, msg_id), &message.encode());
-}
-
-pub(crate) fn get_waiting_message(prog_id: H256, msg_id: H256) -> Option<Message> {
-    sp_io::storage::get(&wait_key(prog_id, msg_id))
-        .as_ref()
-        .map(|val| Message::decode(&mut &val[..]).ok())
-        .flatten()
+pub fn insert_waiting_message(prog_id: H256, msg_id: H256, message: Message, bn: u32) {
+    let payload = (message, bn);
+    sp_io::storage::set(&wait_key(prog_id, msg_id), &payload.encode());
 }
 
-pub fn remove_waiting_message(prog_id: H256, msg_id: H256) -> Option<Message> {
+pub fn remove_waiting_message(prog_id: H256, msg_id: H256) -> Option<(Message, u32)> {
     let id = wait_key(prog_id, msg_id);
-    let msg: Option<Message> = sp_io::storage::get(&id)
-        .map(|val| Message::decode(&mut &val[..]).expect("message encoded correctly"));
+    let msg = sp_io::storage::get(&id)
+        .map(|val| <(Message, u32)>::decode(&mut &val[..]).ok())
+        .flatten();
```

Comment:
Grumble: `map` + `flatten` may be replaced with `and_then`.
```suggestion
        .and_then(|val| <(Message, u32)>::decode(&mut &val[..]).ok());
```